src/profile-logic/processed-profile-versioning.js

```diff
@@ -2390,6 +2390,57 @@ const _upgraders = {
     // profiles from Firefox use subcategories to represent the information
     // about the JIT type of a JS frame.
 
+    // Very old Gecko profiles don't have JS subcategories. Convert the
+    // implementation information to subcategories.
+    function maybeConvertImplementationToSubcategories(profile) {
+      const { categories } = profile.meta;
+      if (!categories) {
+        return;
+      }
+
+      if (categories.some((c) => c.subcategories.length !== 1)) {
+        // This profile has subcategories.
+        return;
+      }
+
+      const jsCategoryIndex = categories.findIndex(
+        (c) => c.name === 'JavaScript'
+      );
+      if (jsCategoryIndex === -1) {
+        // This profile has no JavaScript category.
+        return;
+      }
+
+      const jsCategorySubcategories = categories[jsCategoryIndex].subcategories;
+      const subcategoryForImplStr = new Map();
+
+      for (const thread of profile.threads) {
+        const subcategoryForImplStrIndex = new Map();
```

Comment:
I've been wondering if this 1st level of caching was really useful but I also don't mind.
(I guess it's saving a stringArray lookup + probably the map operation with an integer is faster?)

Reply:
done